reactor-core/src/main/java/reactor/core/publisher/Operators.java

```diff
@@ -1479,7 +1506,190 @@ public Context currentContext() {
+	 * {@link Schedulers#newElastic(String) elastic Scheduler}.
+	 * <p>
+	 * The desired maximum stack depth (or rather the maximum desired operator chain depth) is
+	 * defined via the {@code reactor.max.operator.depth} system property.
+	 */
+	static final class Stacksafe {
+
+		final long maxDepth;
+		long totalDepth;
+		long depthSinceLastProtect;
+
+		Stacksafe() {
+			this(Operators.stacksafeMaxOperatorDepth);
+		}
+
+		Stacksafe(long maxDepth) {
+			this.maxDepth = maxDepth;
+			this.totalDepth = 0L;
+			this.depthSinceLastProtect = 0L;
+		}
+
+		Scheduler.Worker createWorker() {
+			Scheduler s = stacksafeSchedulerRef.get();
+			if (s != null) {
+				return s.createWorker();
+			}
+
+			Scheduler stacksafeScheduler =
+					Schedulers.newParallel(Schedulers.DEFAULT_POOL_SIZE,
+							Schedulers.createNonBlockingThreadFactory("stacksafe",
+									COUNTER,
+									true));
+
+			if (stacksafeSchedulerRef.compareAndSet(null, stacksafeScheduler)) {
+				s = stacksafeScheduler;
+			}
+			else {
+				stacksafeScheduler.dispose();
+				s = stacksafeSchedulerRef.get();
+			}
+			return s.createWorker();
+		}
+
+		<U> CoreSubscriber<U> protect(CoreSubscriber<U> actual,
+				OptimizableOperator<?, ?> operator) {
+			if (maxDepth <= 0) {
+				return actual;
+			}
+
+			totalDepth++;
+			depthSinceLastProtect++;
+			CoreSubscriber<U> result = actual;
+			//reset the perceived stack depth when encountering thread-hopping operators
+			if (actual instanceof FluxPublishOn.PublishOnSubscriber || actual instanceof MonoPublishOn.PublishOnSubscriber || actual instanceof FluxSubscribeOn.SubscribeOnSubscriber || actual instanceof MonoSubscribeOn.SubscribeOnSubscriber) {
+				depthSinceLastProtect = 0;
+				return actual; //don't trampoline right before an explicit thread-hopping
+			}
+			if (depthSinceLastProtect == maxDepth) {
+				depthSinceLastProtect = 0;
+				Scheduler.Worker worker = createWorker();
+
+				if (actual instanceof QueueSubscription) {
+					//original was a QueueSubscription, emulate one (that always negotiate NONE)
+					result = new FluxHide.SuppressFuseableSubscriber<>(actual);
+					result = new StacksafeSubscriber<>(result,
+							worker,
+							"depth=" + totalDepth + ", wrapping=" + operator);
+				}
+				else {
+					result = new StacksafeSubscriber<>(actual,
+							worker,
+							"depth=" + totalDepth + ", wrapping=" + operator);
+				}
+			}
+			return result;
+		}
+
+		static final class StacksafeSubscriber<T> implements InnerOperator<T, T> {
+
+			final CoreSubscriber<? super T> downstream;
+			final Scheduler.Worker          worker;
+			final String                    stringRepresentation;
+
+			Subscription subscription;
+
+			StacksafeSubscriber(CoreSubscriber<? super T> downstream,
+					Scheduler.Worker worker, String message) {
+				this.worker = worker;
+				this.downstream = downstream;
+				this.stringRepresentation = "StacksafeSubscriber{" + message + "}";
```

Comment:
`reactor.core.CoreSubscriber#currentContext` is not taken into account  
need to cache it here

Reply:
```java
@Test
void largeOperatorChainSmallerTrampoliningThresholdTrampolines() {
	Set<String> threadNames = new ConcurrentSkipListSet<>();
	Flux<Integer> tooDeep = Flux.deferWithContext(ctx -> Flux.just(0).hide());
	Flux<Integer> tooDeep = Flux.just(0).hide();
	int bound = 7_000;
	for (int i = 0; i <= bound; i++) {
		int currentI = i;
		tooDeep = tooDeep
				.doOnSubscribe(s -> threadNames.add(Thread.currentThread().getName()))
				.map(previous -> currentI);
	}
	assertThat(tooDeep.blockLast(Duration.ofSeconds(5))).isEqualTo(bound);
	assertThat(threadNames).startsWith(Thread.currentThread().getName());
	assertThat(threadNames.size()).as("number of threads").isGreaterThan(1);
}
```
produces 
```
java.lang.StackOverflowError
	at reactor.core.publisher.FluxMap$MapSubscriber.actual(FluxMap.java)
	at reactor.core.publisher.InnerOperator.currentContext(InnerOperator.java:32)
	at reactor.core.publisher.FluxPeek$PeekSubscriber.currentContext(FluxPeek.java:112)
```